### ISSUE TITLE
Doxygen new TOC implementation produces invalid XML output

### ISSUE THREAD

**Issue by mosra** [mosra/m.css#73]
Since doxygen/doxygen#736 Doxygen generates a full TOC table for the XML output instead of just a placeholder (which I implemented before). The m.css theme doesn't use that (and doesn't plan to be, since our existing implementation is good enough and we generate the TOC for other things anyway), but the problem is that Doxygen generates invalid XML for multiple `@section`s following each other -- in that case, the `</tocsect>` isn't inserted and there's lots of open `<tocsect>` that are never closed. The Doxygen test doesn't verify such case at all, only nested sections/subsections/subsubsections.

Code in question is https://github.com/doxygen/doxygen/pull/736/files#diff-dea13e25f9bbae3ddde44cc615b9c9ba. I can't fathom *why* it has to insert a redundant `<tableofcontents>` as well, nevertheless that's just a minor thing compared to this.

Repro case is in https://gist.github.com/jbakosi/502091bddf05777b77dc684b363c05ba#file-gistfile1-txt-L28 (thanks, @jbakosi)

Possible workarounds until this is fixed:

- don't use more than one `@section` following each other, only nest them (hahahah)
- don't use `@tableofcontents` at all
- use https://github.com/mosra/doxygen/tree/mosra-master or a commit before the aforementioned PR got merged

TODO:

- [x] ~~minimize a repro case~~ not a problem since https://github.com/doxygen/doxygen/commit/185d6abdc832e7dd66183a2154a13a546414b96f
- [x] ~~update their test and expected output~~ it doesn't test properly, but i don't care
- [x] ~~if brave enough, try to fix their code (and get rid of the nested `<tableofcontents>` atrocity)~~ i don't care
- [x] ~~submit a PR~~
- [x] ~~wait for it to get fixed/merged~~

As with [all other issues](https://github.com/mosra/m.css/labels/3rd%20party) that require submitting PRs/code to Doxygen, any help with getting it to behave properly is appreciated :)

**Comment by mosra** [comment]
Release of Doxygen 1.8.15 is [scheduled for the end of this year](https://github.com/doxygen/doxygen/issues/6393#issuecomment-448165197) and it's crucial to have this reported and fixed before the release is out, otherwise I'd need to further diverge with my forked version. I hope I'm able to submit a PR with a fix until then. If anybody is wiling to help with this, I'd be very happy.

**Comment by mosra** [comment]
@jbakosi I tested with current Doxygen master and I'm not able to reproduce this when enabling "the new TOC" on various pages -- the generated XMLs are valid and m.css processes them without problems. Could it be that https://github.com/doxygen/doxygen/commit/185d6abdc832e7dd66183a2154a13a546414b96f *by a complete accident* (ugh) fixed this issue and you had tested with a build from before this was integrated?

Besides that, I also opened https://github.com/doxygen/doxygen/issues/6692 to hopefully fix the annoying backwards incompatibility.

**Comment by jbakosi** [comment]
Yep, [doxygen@449a7e2b](https://github.com/doxygen/doxygen/commit/449a7e2b4ff114a72be573013558bae19672ebbc) appears to beahve (almost) correctly regarding TOC. Almost, because the TOC does not appear at the position on the page where the `@tableofcontents{xml}` command is inserted but always after the title. This is only mildly annoying, so I guess it is progress?

**Comment by jbakosi** [comment]
Okay, I'm a little confused. I just re-tested [doxygen@9316a086](https://github.com/doxygen/doxygen/commit/449a7e2b4ff114a72be573013558bae19672ebbc), which is what I have been using for a little while and that also generates the TOC correctly. I'm not sure when the TOC was fixed...

**Comment by mosra** [comment]
Both https://github.com/doxygen/doxygen/commit/9316a08624d834a635a82b5aa394ef621dbf513d (Aug 15) and https://github.com/doxygen/doxygen/commit/449a7e2b4ff114a72be573013558bae19672ebbc (Dec 18) are after https://github.com/doxygen/doxygen/commit/185d6abdc832e7dd66183a2154a13a546414b96f (Jul 22), which (I think) fixed the issue, so it makes sense :)

**Comment by jbakosi** [comment]
Okay, top-TOC is still better than no TOC. Thanks for the explanation. Though, if I remember correctly, the TOC wherever the command is worked a few months ago.

**Comment by mosra** [comment]
> Though, if I remember correctly, the TOC wherever the command is worked a few months ago.

Nope, it was always put after the brief docs and before detailed docs by the theme. Doxygen extracts the `@tableofcontents` from wherever it is inside the text and converts it to a property of the page, losing the position information completely. This worked only in the m.css Pelican theme, but never for Doxygen ;)

I opened #77 with some ideas how this could be done. https://github.com/doxygen/doxygen/issues/6692 is also fixed now (and released in 1.8.15), so you don't need to say `{xml}` anymore (I find it absolutely redundant myself).

Testing myself, I didn't spot any issues related to broken XML output with the 1.8.15 releases (but got lots of other breakages, of course), so I think this can be closed :)

**Comment by jbakosi** [comment]
Ah okay. That is what I remembered: the brief doc, TOC, then contents. Thanks for the clarification. I will try removing `{xml}`.

**Comment by mosra** [comment]
Ugh. Testing further, I'm getting randomly truncated or completely empty XML files after https://github.com/doxygen/doxygen/issues/6692 (or maybe due to the new TOC implementation). So nope, nothing got fixed, the dumpster fire burns on. 

Reopening until I'm sure what the hell happened.
